### ISSUE TITLE
Referencing href property of an link or anchor that has just scheme in the href causes an exception

### ISSUE THREAD

**Issue by nlysaght** [AngleSharp/AngleSharp#341]
Referencing the **href** property of a link element `IHtmlLinkElement` or `IHtmlAnchorElement` where the **href** contains only the scheme causes an exception.
Examples
This link would cause an exception
`<link rel="shortcut icon" href="https:" />`
Likewise the following anchor would cause an exception.
`<a href="http:\">On The Park</a>`

Unfortunately it would appear that a lot of sites have both links and anchors implemented this way for some reason I can't quite understand.

It would appear that the AngleSharp.Url.ParseScheme is the cause, both exception trace back to it in the call stack.
`
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at AngleSharp.Url.ParseScheme(String input, Boolean onlyScheme)
   at AngleSharp.Url.ParseUrl(String input, Url baseUrl)
   at AngleSharp.Extensions.ElementExtensions.GetUrlAttribute(Element element, String name)
   at AngleSharp.Dom.Html.HtmlLinkElement.get_Href()
`

**Comment by nlysaght** [comment]
I'll take a stab at fixing this myself over the weekend, and will hopefully revert back with a pull request soon.

**Comment by FlorianRappl** [comment]
That would be awesome! Thanks a lot for the report and eagerly waiting for the PR to have this fix in v0.9.7

**Comment by nlysaght** [comment]
Hi @FlorianRappl , the fix for this is relatively straight forward.
Line 494 of the Url core, needs this line, placed before the `if(input[++index....`
`if (index < input.Length-1)`.
The more complicated part of this is around the unit testing, I'm unsure if we consider a Url with just the scheme as valid.
For example, this test passes

```
          {
                var address = "http";
                var result = new Url(address);
                Assert.IsFalse(result.IsInvalid);
           }

```

But this test does not pass

```
        {
                var address = "http://";
                var result = new Url(address);
                Assert.IsFalse(result.IsInvalid);
            }
```

So the basic question comes down to is a Scheme only Url a valid Url?

**Comment by FlorianRappl** [comment]
I would not consider a scheme-only URL a valid URL - I mean there is strictly speaking no syntax error, however, there is essentially no URL given at all.

We only know the scheme, but nothing else (hostname, path, ...), i.e., we cannot use the URL to navigate somewhere. I would therefore mark it as invalid to hint the missing part (at least a hostname is required in my opinion).

Thanks for taking care of this @nlysaght! :+1:

**Comment by nlysaght** [comment]
Hi @FlorianRappl, This is getting a bit cumbersome to implement, basically before we parse we need to validate if we have a valid scheme, once we determine it is an absolute Url (!_relative).
There are various routines for fixing up broken or incorrectly formatted url's. Checking for the absence of a scheme or an incorrectly formatted scheme before these can break these tests. For example to validate we have a scheme we must find something before a **://**, if we do this it would break these tests like **DocumentUrlAddMissingSlashAfterProtocolForFile**, or **DocumentUrlAddMissingSlashAfterProtocolForFtp**.  
Would you have an opinion on an approach for this? Maybe managing failures in Url formats should not be tied in with the Url class itself, it would seem to me that these would be a separate concern or responsibility where the actual url itself has failed? But without understanding the use cases for fixing up the Url's in the first place it hard to define the correct solution for this.

**Comment by FlorianRappl** [comment]
If its easier to implement when scheme-only URLs are considered valid; let's go for this approach. As described the URL is valid from a syntax-POV so I see no problem.

No matter how we implement it - the existing tests need to pass. Some of the URL tests are taken from large independent URI verification test suites, i.e., we have to cover them to be standard conform.

Regarding a fix I can say that I wouldn't go so far. There is no way we can _reliably_ fix broken URLs. Hence I would not even try.

**Comment by nlysaght** [comment]
I think the easiest & least intrusive fix, is to just put in the original 1 line fix to stop the crash from happening, everything after that works as expected. There is some confusion, e.g. a url of **http** is considered valid because it's a relative url to a resource called http, but a url of **http://** is considered invalid as there is no path or authority associated with it . But these are such edge cases I think it hardly matters, the main issue here is stopping the code from causing an exception. Would you agree?

**Comment by FlorianRappl** [comment]
I definitely agree. I would say such ambiguous URLs are hard to handle correct and hence our priority should be in the robustness sector here. Thanks!

**Comment by nlysaght** [comment]
Hi Florian, I'm attaching a 341.txt as a diff
[341.txt](https://github.com/AngleSharp/AngleSharp/files/266319/341.txt)
 for this issue, unfortunately I'm restricted from adding mono to my environment, so can't do a full build. I've some unit tests in there also, I've tested this internally myself and am happy that all is working as it should and all other tests are passing. Hope this helps and thanks for the great work you have started here. If I find other issues I will try and fix as much as possible myself.

Kind Regards
Noel

**Comment by FlorianRappl** [comment]
Thanks - I just applied the diff! Much appreciated :+1:

**Comment by FlorianRappl** [comment]
Fixed, see PR #344.
